File: Algorithms/C++/01189-Maximum Number of Balloons.cpp
class Solution {
public:
    int maxNumberOfBalloons(string text) {
        std::unordered_map<char, int> map{
            {'b', 0},
            {'a', 0},
            {'n', 0},
            {'l', 0},
            {'o', 0}
        };

        for(int i{0}; i < text.size(); i++){
            if(text[i] == 'b' || text[i] == 'a' || text[i] == 'l' || text[i] == 'o' || text[i] == 'n'){
                map[text[i]]++;
            }
        } 

        int maxBalloons{INT_MAX};
        for(const auto& [key, value] : map){
            if(key == 'b' || key == 'a' || key == 'n'){
                maxBalloons = std::min(maxBalloons, value);
            } else if(key == 'l' || key == 'o'){
                maxBalloons = std::min(maxBalloons, (value/2));
            }
        }
        if(maxBalloons == INT_MAX) return 0;
        return maxBalloons;
    }
};

/*
    in balloon:

    b: 1
    a: 1
    l: 2
    o: 2
    n: 1

    make an unordered map ONLY with these characters
    take the minimum of each char, with chars "l" and "o" divided by 2

    return minimum

*/

File: Algorithms/Python3/01189-Maximum Number of Balloons.py
class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        balloonMap = {
            "b": 1,
            "a": 1,
            "l": 2,
            "o": 2,
            "n": 1
        }
        textMap = {            
            "b": 0,
            "a": 0,
            "l": 0,
            "o": 0,
            "n": 0
        }

        for c in text:
            if c in balloonMap:
                textMap[c] += 1
        
        maxBalloons = int(sys.maxsize)
        for key, value in textMap.items():
            maxBalloons = min(maxBalloons, int(value / balloonMap[key]))
        
        return maxBalloons

'''
    tc: O(n)
    sc: O(n)

    hash map
    b: 1
    a: 1
    l: 2
    o: 2
    n: 1

    make map for balloon
    iterate through our text string to only check chars in "balloon"
    find max number of balloons able to be created
    (minimum multiplier)

'''